### DISCUSSION TITLE
Is it possible to tell if getStaticProps is running as part of the build or revalidation?

### DISCUSSION THREAD

**Discussion by Mitsunee** [vercel/next.js#47845] [Help] [answered]
### Summary

As write permissions are not granted on vercel being able to tell if you are in the build process or not is important information. Is there an enviroment variable or something that I could use to determine whether I'm currently in the build process or in revalidation?

### Example

```js
export async function getStaticProps() {
  if ( whatToUseHere ) {
    // do some build stuff and write files
  } else {
    // we're in ISR, so don't write files
  }

  return { props: someData, revalidate: 3600 }
}
```

**Comment by icyJoseph** [comment, marked as answer]
I think you can use `process.env.NEXT_PHASE`, it'll be `phase-production-build`, but only during build time. 

Here are the available phases: https://github.com/vercel/next.js/blob/5e6b008b561caf2710ab7be63320a3d549474a5b/packages/next/shared/lib/constants.ts#L19-L23

Use them like:

```js
import { PHASE_PRODUCTION_BUILD } from "next/constants";

export async function getStaticProps() {
  if (process.env.NEXT_PHASE === PHASE_PRODUCTION_BUILD ) {
    // do some build stuff and write files
  } else {
    // we're in ISR, so don't write files
  }

  return { props: someData, revalidate: 3600 }
}
```

Alternatively you could inject your own env vars from `next.config.js`, based on the phase as well: https://nextjs.org/docs/api-reference/next.config.js/introduction

In `start` mode I think this will be undefined though ~~, so you can't use it there, then you'd need to inject your own env vars from `next.config.js`.

**Reply by Mitsunee** [reply]
That looks promising, doesn't really matter that it'll be `undefined` during ISR because the check should still work as intended is this case. Might be nice to turn it into a util like

```ts
export function inBuildPhase(orDev?: boolean) {
  if (process.env.NEXT_PHASE == PHASE_PRODUCTION_BUILD) return true;
  if (process.env.NEXT_PHASE == PHASE_DEVELOPMENT_SERVER) return orDev ?? true;
  return false;
}
```

That should then return `true` during build (and potentially dev) only, right?

**Reply by icyJoseph** [reply]
_I think so_, It is a bit scary that these are not documented beyond the `next.config.js` scope, but it is what it is... If it does work out, please don't forget to mark as answered.

**Reply by Mitsunee** [reply]
`process.env.NEXT_PHASE` seems to be `undefined` during dev, but technically this still does answer the original question. Might have to poke around more to see if/where the `PHASE_DEVELOPMENT_SERVER` is used another day.

**Comment by bfellows37** [comment]
helpful discussion :+1:
